SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,203 @@
+            if (result != raw.SQLITE_DONE)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to run query, result={result}\n\t{errorMessage}");
+            }
+            return raw.sqlite3_finalize(stmt);
         }
 
-        public Task<int> InsertAsync<T>(T val) where T : new()
+        private List<object[]> ExecuteSelectionSqlQuery(string query, IList<object> args = null)
         {
-            try
+            var db = _db ?? throw new StorageException("The database wasn't initialized.");
+            var entries = new List<object[]>();
+            var queryResult = raw.sqlite3_prepare_v2(db, query, out var stmt);
+            if (queryResult != raw.SQLITE_OK)
             {
-                return _dbConnection.InsertAsync(val);
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to prepare SQL query, result={queryResult}\n\t{errorMessage}");
             }
-            catch (SQLiteException e)
+            BindParameters(stmt, args);
+            while (raw.sqlite3_step(stmt) == raw.SQLITE_ROW)
             {
-                throw ToStorageException(e);
+                var count = raw.sqlite3_column_count(stmt);
+                entries.Add(Enumerable.Range(0, count).Select(i => GetColumnValue(stmt, i)).ToArray());
             }
+            var result = raw.sqlite3_finalize(stmt);
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to finalize SQL query, result={result}\n\t{errorMessage}");
+            }
+            return entries;
         }
 
-        private static StorageException ToStorageException(SQLiteException e)
+        public void CreateTable(string tableName, string[] columnNames, string[] columnTypes)
         {
-            return new StorageException($"SQLite errorCode={e.Result}", e);
+            var tableClause = string.Join(",", Enumerable.Range(0, columnNames.Length).Select(i => $"{columnNames[i]} {columnTypes[i]}"));
+            var result = ExecuteNonSelectionSqlQuery($"CREATE TABLE IF NOT EXISTS {tableName} ({tableClause});");
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to create table, result={result}\n\t{errorMessage}");
+            }
         }
 
-        public async Task<int> DeleteAsync<T>(Expression<Func<T, bool>> pred) where T : new()
+        public int Count(string tableName, string columnName, object value)
         {
-            try
+            var result = ExecuteSelectionSqlQuery($"SELECT COUNT(*) FROM {tableName} WHERE {columnName} = ?;", new[] { value });
+            var count = (long)(result.FirstOrDefault()?.FirstOrDefault() ?? 0L);
+            return (int)count;
```

Comment:
Why not using long all the way through and modify interface?

Reply:
We thought about it, but it leads to modify interface if storage too. Cast here to minimize the changes. Anyway, int more than enough here